Add support for extended colors.

auouymous opened the pull request:
The original code made no attempt to clamp the upper limit of colors, allowing normal extended color values to address beyond the `color_pair_table` array. This fixes a segfault when using large background numbers such as 255.

The new code allows any color value supported by the terminal, clamping the value and logging an error for unsupported colors.

brasko commented:
Thanks for submitting this. I'll take a look at this once when I get some time, before the next release.

auouymous commented:
The new hl_get_color_attr_from_index() queries color pairs for what seems like every terminal cell on any update. This requires color pairs to be stored in a 129kB array while still initializing them as needed to avoid exceeding the limit.

auouymous commented:
It fixes the segfault and adds support for 16, 88 and 256 color terminals

The current 9x9 array only supports 8 colors, and any fg value over 81 or any bg value over 9 will access outside the array. Setting a bg color to 255 always generates a segfault for me.

But it isn't as simple as increasing the array size to handle more colors because there are probably only 32k color pairs available for 256 colors and 257x257 requires double that. So the code creates color pairs as needed, with most of the array containing unset values.

The new code accepts 0-255 values if the terminal supports 256 colors, or 0-87 if 88 colors or 0-15 if 16 colors, and so on for the number of colors supported by the given terminal. I used a static array but it should probably allocate the array for the given terminal so it uses less memory when fewer colors are supported. A dynamic array would allow more than 256 colors, but we probably don't want to do that because 16 million colors would allocate more than 256TB.

I do not get any errors, what are they?

auouymous commented:
The following works fine on a 256 terminal with this PR.

```
hi StatusLine ctermfg=251 ctermbg=237 cterm=bold
hi ScrollModeStatus ctermfg=251 ctermbg=237 cterm=bold
```

But libvterm is only asking for 0-7 when running cgdbutil_driver on a 256 color terminal. The problem is that `get_ncurses_color_index()` mangles the colors because cgdbutil_driver displays fine with the following patch.

```patch
diff --git a/cgdb/vterminal.cpp b/cgdb/vterminal.cpp
index e03c635..91cd7a8 100644
--- a/cgdb/vterminal.cpp
+++ b/cgdb/vterminal.cpp
@@ -330,7 +330,7 @@ static int get_ncurses_color_index(VTermColor &color, bool &bold)
     } else if (VTERM_COLOR_IS_DEFAULT_BG(&color)) {
         index = -1;
     } else if (VTERM_COLOR_IS_INDEXED(&color)) {
-        if (color.indexed.idx >= 0 && color.indexed.idx < 16) {
+        if (color.indexed.idx >= 0 && color.indexed.idx < 256) {
             index = color.indexed.idx;
         } else if (color.indexed.idx >=232) {
             int num = color.indexed.idx;
```

This PR is kind of working as intended. cgdb allocates a bunch of default color pairs and cgdbutil_driver then uses 64 colors, which exceeds the 64 color pairs ncurses reports. But unpatched cgdb uses 81 color pairs and swin_init_pair() doesn't log an error. So it looks like ncurses actually supports reported color pairs + (reported colors * 2 + 1) when default color is supported.

However, the patch would throw an error every time the same color was requested, which libvterm guarantees to happen, a lot. I updated it to add color pairs until swin_init_pair() fails, and then store 0 in the array so only the first use of each color combination will fail. I also fixed it to store 0 for the -1,-1 pair instead of allocating a new color pair for it. And I allocated the array to reduce memory usage on low color terminals.
